keps/sig-api-machinery/3396-websocket-protocol-enhancements/README.md

```diff
@@ -0,0 +1,869 @@
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Protocol changes
+
+Currently, API server supports [binary and base64 WebSocket protocols](https://github.com/kubernetes/kubernetes/blob/3ddd0f45aa91e2f30c70734b175631bec5b5825a/staging/src/k8s.io/apiserver/pkg/util/wsstream/conn.go#L34-L64). 
+
+Binary protocol: channel id `0xFF` can be used as a control channel. First byte of the message contents can be used
+to encode the command and the rest of the message can be used to encode the payload for the command.
+
+Binary protocol control commands:
+
+- `0x00`: half-close the channel. Next byte is the channel id.
+- `0x01`: reset the channel. Next byte is the channel id.
+
+Base64-based protocol: can use character `c` (for `control`) as the channel id for the control channel.
+
+Base64-based protocol control commands:
+
+- `h`: half-close the channel. Next character is the channel id.
+- `r`: reset the channel. Next character is the channel id.
+
+### Protocol negotiation
+
+If we have a new protocol, then older servers will not have support for it. In that case a newer client will try
+v5 but an older server will not accept that. To be backwards compatible, ideally client should use the intended
+mechanism that HTTP `Upgrade` provides - send a list of protocols. I.e. client would send a header with
+something like `Upgrade: v5.channel.k8s.io, SPDY/3.1` and server would pick a protocol it supports,
+taking into account client's preference (the list is ordered).
+
+The difficulty with the above is that both Gorilla WebSocket and SPDY libraries are not built to allow to
+use them this way. Both encapsulate negotiation, and you cannot compose them to get the above behavior. 
+Refactoring them is not an option since SPDY is dead and nobody will spend time on the library and Gorilla
+WebSocket is [looking for maintainers](https://github.com/gorilla/websocket/issues/370) and hence
+a PR is unlikely to get merged quickly even if at all (since this is a big API addition/change).
+
+An alternative is to try to establish `v5.channel.k8s.io` and fall back to `SPDY/3.1` but that is 2x
+the number of requests when newer client is used with an older server. Perhaps this is acceptable
+since `exec`, `attach`, and `cp` are I/O-heavy anyway and are not used all the time. This is a temporary
+overhead, eventually this situation will become less and less frequent as servers upgrade to a version with
+the new protocol. If the overhead is not acceptable, user can almost always use a client of a matching version.
+At some point SPDY support should be removed and this problem will go away completely.
```

Comment:
Consider moving this under `## Alternatives`, and replace the moved text with a link to it.

Reply:
This is an alternative to the protocol negotiation section only, not to the whole proposal. I think it will be out of context if move to another section.